pkg/controller/endpointslicemirroring/reconciler.go

```diff
@@ -0,0 +1,405 @@
+				slicesToDelete = append(slicesToDelete, existingSlice)
+			}
+		}
+	}
+
+	mirroringmetrics.EndpointsAddedPerSync.WithLabelValues().Observe(float64(totalAdded))
+	mirroringmetrics.EndpointsRemovedPerSync.WithLabelValues().Observe(float64(totalRemoved))
+
+	endpointsNN := types.NamespacedName{Name: endpoints.Name, Namespace: endpoints.Namespace}
+	r.metricsCache.UpdateEndpointPortCache(endpointsNN, epMetrics)
+
+	return r.finalize(endpoints, slicesToCreate, slicesToUpdate, slicesToDelete)
+}
+
+// reconcileByPortMapping compares the endpoints found in existing slices with
+// the list of desired endpoints and returns lists of slices to create, update,
+// and delete. The logic is split up into several main steps:
+// 1. Iterate through existing slices, delete endpoints that are no longer
+//    desired and update matching endpoints that have changed.
+// 2. Iterate through slices that have been modified in 1 and fill them up with
+//    any remaining desired endpoints.
+// 3. If there still desired endpoints left, try to fit them into a previously
+//    unchanged slice and/or create new ones.
+func (r *reconciler) reconcileByPortMapping(
+	endpoints *corev1.Endpoints,
+	existingSlices []*discovery.EndpointSlice,
+	desiredSet endpointSet,
+	endpointPorts []discovery.EndpointPort,
+	addressType discovery.AddressType,
+) ([]*discovery.EndpointSlice, []*discovery.EndpointSlice, []*discovery.EndpointSlice, int, int) {
+	slicesByName := map[string]*discovery.EndpointSlice{}
+	sliceNamesUnchanged := sets.String{}
+	sliceNamesToUpdate := sets.String{}
+	sliceNamesToDelete := sets.String{}
+	numRemoved := 0
+
+	// 1. Iterate through existing slices to delete endpoints no longer desired
+	//    and update endpoints that have changed
+	for _, existingSlice := range existingSlices {
+		slicesByName[existingSlice.Name] = existingSlice
+		newEndpoints := []discovery.Endpoint{}
+		endpointUpdated := false
+		for _, endpoint := range existingSlice.Endpoints {
+			got := desiredSet.Get(&endpoint)
+			// If endpoint is desired add it to list of endpoints to keep.
+			if got != nil {
+				newEndpoints = append(newEndpoints, *got)
+				// If existing version of endpoint doesn't match desired version
+				// set endpointUpdated to ensure endpoint changes are persisted.
+				if !endpointsEqualBeyondHash(got, &endpoint) {
+					endpointUpdated = true
+				}
+				// once an endpoint has been placed/found in a slice, it no
+				// longer needs to be handled
+				desiredSet.Delete(&endpoint)
+			}
+		}
+
+		// If an endpoint was updated or removed, mark for update or delete
+		if endpointUpdated || len(existingSlice.Endpoints) != len(newEndpoints) {
+			if len(existingSlice.Endpoints) > len(newEndpoints) {
+				numRemoved += len(existingSlice.Endpoints) - len(newEndpoints)
+			}
+			if len(newEndpoints) == 0 {
+				// if no endpoints desired in this slice, mark for deletion
+				sliceNamesToDelete.Insert(existingSlice.Name)
```

Comment:
can we take care of this at the very end somehow (e.g. loop through all slices and those that end up with no endpoints will be marked for deletion)

Reply:
Potentially resolved with last commit - see comment above for more context.